RecoEgamma/EgammaElectronProducers/plugins/alpaka/ElectronNHitSeedAlpakaProducer.cc

```diff
@@ -0,0 +1,227 @@
+#include "HeterogeneousCore/AlpakaCore/interface/alpaka/stream/SynchronizingEDProducer.h"
+
+// Additional includes for testing / comparing implementations
+#include "PixelMatchingAlgo.h"
+
+using Vector3d = Eigen::Matrix<double, 3, 1>;
+
+namespace ALPAKA_ACCELERATOR_NAMESPACE {
+
+	class ElectronNHitSeedAlpakaProducer : public global::EDProducer<> {
+		public:
+		ElectronNHitSeedAlpakaProducer(const edm::ParameterSet& pset):
+        	EDProducer(pset),
+			deviceToken_{produces()},
+			size_{pset.getParameter<int32_t>("size")},
+			initialSeedsToken_(consumes(pset.getParameter<edm::InputTag>("initialSeeds"))),
+			beamSpotToken_(consumes(pset.getParameter<edm::InputTag>("beamSpot"))),
+			magFieldToken_(esConsumes()),
+			geomToken(esConsumes())
+			{
+				superClustersTokens_ = consumes(pset.getParameter<edm::InputTag>("superClusters"));
+			}
+
+      	void produce(edm::StreamID sid, device::Event& event, device::EventSetup const& iSetup) const override {
+		//void produce(device::Event& event, device::EventSetup const& iSetup) const override {
+
+			auto vprim_ = event.get(beamSpotToken_).position();
+			GlobalPoint vprim(vprim_.x(), vprim_.y(), vprim_.z());
+			Vector3d vertex{vprim.x(),vprim.y(),vprim.z()};
+
+			// NEW EVENT
+			std::cout<<" -----> NEW EVENT <------- "<<std::endl;
+			// Get MagField ESProduct for comparing & Geom ESProduct
+			// auto const& magField = iSetup.getData(magFieldToken_);
+			// const TrackerGeometry* theG = &iSetup.getData(geomToken);
+			// PropagatorWithMaterial backwardPropagator_ = PropagatorWithMaterial(oppositeToMomentum, 0.000511, &magField);
+
+			std::vector<reco::SuperClusterRef> superClusterRefVec;
+
+			for (auto& superClusRef : event.get(superClustersTokens_))
+				superClusterRefVec.push_back(superClusRef);
+
+			int32_t superClusterCollectionSize = superClusterRefVec.size();
+			reco::SuperclusterHostCollection hostProductSCs{superClusterCollectionSize, event.queue()};
+
+			std::vector<TrajectorySeed> seedRefVec;
+			for (auto& initialSeedRef : event.get(initialSeedsToken_))
+				seedRefVec.push_back(initialSeedRef);
+
+			int32_t seedCollectionSize = seedRefVec.size();
+			reco::EleSeedHostCollection hostProductSeeds{seedCollectionSize, event.queue()};
+
+			auto& viewSCs = hostProductSCs.view();
+			auto& viewSeeds = hostProductSeeds.view();
+
+			std::cout<<" -----> Collection sizes SCs: "<<superClusterCollectionSize << " & Seeds " <<seedCollectionSize<<std::endl;
+
+			////////////////////////////////////////////////////////////////////
+			// Fill in SOAs
+			// Technically should separate in different producers that create the SoAs ?
+			// Info on SoAs : https://github.com/cms-sw/cmssw/blob/master/DataFormats/SoATemplate/README.md
+			////////////////////////////////////////////////////////////////////
+
+			int32_t i = 0;
+			for (auto& superClusRef : superClusterRefVec)
+			{
+				viewSCs[i].id() =  i;
+				viewSCs[i].scSeedTheta() =  superClusRef->seed()->position().theta();
```

Comment:
It would be more efficient to de-reference the Ref once
```suggestion
				const auto& superClus = *superClusRef;
				viewSCs[i].scSeedTheta() =  superClus.seed()->position().theta();
```